### PR TITLE
setCustomerIds - Feature/core 26064

### PR DESCRIPTION
**Pull request by jileeshs**
Adding setCustomer IDs command

## Description

Exposing new command which allows customers to send declared IDs to Adobe Experience platform

## Related Issue

https://jira.corp.adobe.com/browse/CORE-26064

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] All tests pass and I've made any necessary test changes.
- [x] I have run the Sandbox successfully.

### CODE REVIEW

File: src/components/Identity/index.js [hunk 1]
```diff
@@ -29,11 +36,50 @@ const createIdentity = ({ config, logger, cookie }) => {
...
+    const event = createEvent();
+    validateCustomerIDs(options.customerIDs);
+    Object.keys(options.customerIDs).forEach(key => {
+      customerIDs[key] = options.customerIDs[key];
+    });
+    const normalizedCustomerIDs = normalizeCustomerIDs(customerIDs);
+    const customerIDsHash = createHashFromString(
+      serializeCustomerIDs(normalizedCustomerIDs)
+    );
+    const hasSynced = customerIDsHash === cookie.get("CIDH");
```
Comment (jonsnyder) [inline src/components/Identity/index.js]: Use the constant you created for CIDH (and on line 72)

File: src/components/Identity/index.js [hunk 2]
```diff
@@ -29,11 +36,50 @@ const createIdentity = ({ config, logger, cookie }) => {
...
+      });
+
+    return responsePromise;
+  };
+
+  const setCustomerIDs = options => {
+    const event = createEvent();
+    validateCustomerIDs(options.customerIDs);
+    Object.keys(options.customerIDs).forEach(key => {
+      customerIDs[key] = options.customerIDs[key];
```
Comment (jonsnyder) [inline src/components/Identity/index.js]: What do you think about cloning the object?

File: src/components/Identity/index.js [hunk 3]
```diff
@@ -29,11 +36,50 @@ const createIdentity = ({ config, logger, cookie }) => {
...
+    });
+    const normalizedCustomerIDs = normalizeCustomerIDs(customerIDs);
+    const customerIDsHash = createHashFromString(
+      serializeCustomerIDs(normalizedCustomerIDs)
+    );
+    const hasSynced = customerIDsHash === cookie.get("CIDH");
+    event.mergeMeta({ identity: { customerIDs, hasSynced } });
+
+    if (!hasSynced) {
+      cookie.set("CIDH", customerIDsHash);
```
Comment (jonsnyder) [inline src/components/Identity/index.js]: Does this mean you cannot change the customer IDs in the cookie?

File: src/components/Identity/index.js [hunk 4]
```diff
@@ -100,6 +146,12 @@ const createIdentity = ({ config, logger, cookie }) => {
     commands: {
       getEcid() {
         return optIn.whenOptedIn().then(getEcid);
+      },
+      setCustomerIDs(options) {
```
Comment (jonsnyder) [inline src/components/Identity/index.js]: Have we considered allowing customers to set the customer ids in the config?  Then the setCustomerIDs data exchange would happen on the first request.

Comment (jonsnyder) [inline src/components/Identity/index.js]: Should this be `setCustomerIds` ?

File: src/components/Identity/index.js (hunk 1 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: Done 👍

File: src/components/Identity/index.js (hunk 2 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: Didn't find an advantage to clone this object here.

File: src/components/Identity/index.js (hunk 3 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: This means we have the same info in the cookie, and hence it was send to identity backend already, there is no need to perform this operation.

File: src/components/Identity/index.js (hunk 4 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: Yes, I think I will do that in a separate task.
Created a subtask here : https://jira.corp.adobe.com/browse/CORE-32507

Comment (jileeshs) [reply src/components/Identity/index.js]: I just used the existing API name in Visitor JS today.

File: src/components/Identity/index.js [hunk 5]
```diff
@@ -12,7 +12,17 @@ governing permissions and limitations under the License.
...
+  serializeCustomerIDs,
+  createHashFromString,
+  validateCustomerIDs,
+  normalizeCustomerIDs
+} from "./identityUtil";
 
+import { COOKIE_NAMES } from "./constants";
+
+const { CUSTOMER_ID_HASH, EXPERIENCE_CLOUD_ID } = COOKIE_NAMES;
 const ECID_NAMESPACE = "ECID";
```
Comment (jfkhoury) [inline src/components/Identity/index.js]: Should we remove this guy?

File: src/components/Identity/index.js [hunk 6]
```diff
@@ -25,15 +35,54 @@ const addIdsContext = (payload, ecid) => {
...
+    validateCustomerIDs(options.customerIDs);
+    Object.keys(options.customerIDs).forEach(key => {
+      customerIDs[key] = options.customerIDs[key];
+    });
+    const normalizedCustomerIDs = normalizeCustomerIDs(customerIDs);
+    const customerIDsHash = createHashFromString(
+      serializeCustomerIDs(normalizedCustomerIDs)
+    );
+    const hasSynced = customerIDsHash === cookie.get(CUSTOMER_ID_HASH);
+    event.mergeMeta({ identity: { customerIDs, hasSynced } });
```
Comment (jfkhoury) [inline src/components/Identity/index.js]: Is that how Konductor expects those? Just double checking that we discussed them with Rares' team.

File: src/components/Identity/index.js [hunk 7]
```diff
@@ -100,6 +149,12 @@ const createIdentity = ({ config, logger, cookie }) => {
     commands: {
       getEcid() {
         return optIn.whenOptedIn().then(getEcid);
+      },
+      setCustomerIDs(options) {
```
Comment (jfkhoury) [inline src/components/Identity/index.js]: Let's rename to `setCustomerIds` since we had this conversation last week about using `d` for configs and commands; for everything basically :)

Comment (jfkhoury) [inline src/components/Identity/index.js]: Should we simplify this command to `setCustomerIds(ids)`?

This way customers will do:

```
alloy("setCustomerIds", {
    email: {
    },
    crm: {
    }
});
```
instead of:

```
alloy("setCustomerIds", {
    customerIds: {
        email: { },
        crm: {}
    }
});
```
@jonsnyder  @Aaronius

File: src/components/Identity/constants.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+const AUTH_STATES = {
```
Comment (jfkhoury) [inline src/components/Identity/constants.js]: Will those be used anywhere other Identity? If so, we should move them with the other top level constants.

File: src/components/Identity/identityUtil.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import { AUTH_STATES } from "./constants";
```
Comment (jfkhoury) [inline src/components/Identity/identityUtil.js]: Should we rename this module to simply `utils.js` since it lives in `Identity` already?

File: src/components/Identity/identityUtil.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import isObject from "../../utils/isObject";
+
+const validateCustomerIDs = customerIDs => {
+  if (!isObject(customerIDs)) {
+    throw new Error(
+      "Invalid customer ID format. Each namespace should be an object"
```
Comment (jfkhoury) [inline src/components/Identity/identityUtil.js]: Let's extract those messages into a single variable to save those precious bits.

File: src/components/Identity/identityUtil.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import { AUTH_STATES } from "./constants";
+import isObject from "../../utils/isObject";
+
+const validateCustomerIDs = customerIDs => {
```
Comment (jfkhoury) [inline src/components/Identity/identityUtil.js]: Let's rename all the `D` to `d`.

File: src/components/Identity/identityUtil.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        "Invalid customer ID format. Each namespace object should have an id"
+      );
+    }
+  });
+};
+const normalizeCustomerIDs = customerIDs => {
+  const normalized = {};
+  Object.keys(customerIDs).forEach(customerID => {
+    const currentID = customerIDs[customerID];
+    normalized[customerID] = { ...currentID };
```
Comment (jfkhoury) [inline src/components/Identity/identityUtil.js]: Does `{ ...currentID }` throw a `Symbol` exception in `IE 11`?

File: src/components/Identity/identityUtil.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    if (
+      currentID.authState !== AUTH_STATES.AUTHENTICATED ||
+      currentID.authState !== AUTH_STATES.LOGGED_OUT
+    ) {
+      normalized[customerID].authState = AUTH_STATES.UNKNOWN;
+    }
+  });
+  return normalized;
+};
+const serializeCustomerIDs = customerIDs => {
```
Comment (jfkhoury) [inline src/components/Identity/identityUtil.js]: Do we need this? Can we simply stringify the JSON instead, then hash that?

File: src/components/Identity/identityUtil.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  return Object.keys(customerIDs)
+    .reduce((serialized, customerID) => {
+      return `${serialized}${customerID}|${customerIDs[customerID].id}${
+        customerIDs[customerID].authState
+      }|`;
+    }, "")
+    .slice(0, -1);
+};
+
+const createHashFromString = string => {
```
Comment (jfkhoury) [inline src/components/Identity/identityUtil.js]: Can we use the crypto APIs here, instead of this helper?

@Aaronius

File: src/components/Identity/index.js [hunk 15]
```diff
@@ -25,15 +35,54 @@ const addIdsContext = (payload, ecid) => {
...
   let deferredForEcid;
+  let network;
+  let lifecycle;
+  const customerIDs = {};
+  const getCustomerIDs = () => customerIDs;
+
+  const makeServerCall = event => {
+    const payload = network.createPayload();
+    payload.addEvent(event);
+    const responsePromise = Promise.resolve()
```
Comment (jfkhoury) [inline src/components/Identity/index.js]: Can't we just return here instead of the `const`?

File: src/components/Identity/index.js [hunk 16]
```diff
@@ -25,15 +35,54 @@ const addIdsContext = (payload, ecid) => {
...
+      })
+      .then(() => {
+        return network.sendRequest(payload, payload.expectsResponse);
+      });
+
+    return responsePromise;
+  };
+
+  const setCustomerIDs = options => {
+    const event = createEvent();
```
Comment (jfkhoury) [inline src/components/Identity/index.js]: I would validate first.

File: src/components/Identity/identityUtil.js (hunk 13 above)
Comment (jileeshs) [reply src/components/Identity/identityUtil.js]: Only advantage is size, Stringify adds a `"` character before and after every key and its value.

File: src/components/Identity/constants.js (hunk 8 above)
Comment (jileeshs) [reply src/components/Identity/constants.js]: I don't know of any other component using these states

File: src/components/Identity/index.js (hunk 6 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: I will double check with Rares in the morning.

File: src/components/Identity/index.js (hunk 16 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: done

File: src/components/Identity/identityUtil.js (hunk 14 above)
Comment (Aaronius) [reply src/components/Identity/identityUtil.js]: Here's how to create a sha-256 hash in a cross-browser way:

```
const text = 'An obscure body in the S-K System, your majesty. The inhabitants refer to it as the planet Earth.';

function buffer2hex(buffer) {
  return Array.prototype.map.call(
    new Uint8Array(buffer),
    function (x) { return ('00' + x.toString(16)).slice(-2) }
  ).join('');
}

function textEncode(str) {
  if (window.TextEncoder) {
    return new TextEncoder('utf-8').encode(str);
  } else {
    // IE 11, which doesn't have TextEncoder
    var utf8 = unescape(encodeURIComponent(str));
    var result = new Uint8Array(utf8.length);
    for (var i = 0; i < utf8.length; i++) {
      result[i] = utf8.charCodeAt(i);
    }
    return result;
  }
}

function digestMessage(message) {
  const data = textEncode(message);
  const crypto = window.msCrypto || window.crypto;
  const result = crypto.subtle.digest('SHA-256', data);
  if (result.then) {
    return result;
  } else {
    // IE 11, whose result is a CryptoOperation object instead of a promise
    return new Promise(function(resolve, reject) {
      result.addEventListener("complete", function() {
        resolve(result.result);
      });
      result.addEventListener("error", function() {
        reject();
      })
    })
  }
}

digestMessage(text).then(function(digestValue) {
  // Should result in 6efd383745a964768989b9df420811abc6e5873f874fc22a76fe9258e020c2e1
  console.log("Digest value in hex", buffer2hex(digestValue));
});
```

Comment (Aaronius) [reply src/components/Identity/identityUtil.js]: I don't fully understand what the hash is used for, but are we sure that SHA-256 is what we want to use vs, say, SHA-1 or SHA-512?

Comment (jfkhoury) [reply src/components/Identity/identityUtil.js]: Thanks @Aaronius.

I think it's used today to hash the customer ids and auth states passed in, and store that hash in the cookie, so if the customers calls us again with the same IDs, we don't try to sync them since we've seen those IDs before.

So I don't think it's necessary to use 256

File: src/components/Identity/util.js [hunk 17]
```diff
@@ -0,0 +1,61 @@
+import { AUTH_STATES } from "./constants";
+import isObject from "../../utils/isObject";
+
+const ERROR_MESSAGE = "Invalid customer Id format.";
```
Comment (Aaronius) [inline src/components/Identity/util.js]: When we're writing in English, let's use `ID`. The capitalization of `*Id` or `id` is just to fit camelcasing rules (as we perceive the rules).

File: src/components/Identity/util.js [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+const NOT_AN_OBJECT_ERROR = "Each namespace should be an object.";
+const NO_ID_ERROR = "Each namespace object should have an id.";
```
Comment (Aaronius) [inline src/components/Identity/util.js]: When we're writing in English, let's use `ID`. The capitalization of `*Id` or `id` is just to fit camelcasing rules (as we perceive the rules).

File: src/components/Identity/identityUtil.js (hunk 13 above)
Comment (Aaronius) [reply src/components/Identity/identityUtil.js]: If we're using SHA to hash the serialized customer IDs, it doesn't matter what size the serialized value is; the SHA hash output will always be the same size regardless. I agree with @jfkhoury that we should use `JSON.stringify()` here, but regardless of what we use, we need to be smarter about how we serialize. Here's why:

```
JSON.stringify({ a: 1, b: 2 }) === JSON.stringify({ b: 2, a: 1 });
```

Conceptually, both objects represent the same thing to the user, but the serialized value, and thus the hashes, will be different. We probably need to break the object down and alphabetize the entries into an array of entries sorted alphanumerically or something like that, such that the output from the two inputs will always be equal. An automated test should be added for this too.

File: src/components/Identity/index.js [hunk 19]
```diff
@@ -12,28 +12,76 @@ governing permissions and limitations under the License.
...
+
+    return responsePromise;
+  };
+
+  const setCustomerIds = ids => {
+    validateCustomerIds(ids);
+    const event = createEvent();
+    Object.keys(ids).forEach(key => {
+      customerIds[key] = ids[key];
+    });
```
Comment (Aaronius) [inline src/components/Identity/index.js]: If I'm reading this right, it looks like if a user calls `setCustomerIds` twice, the customer IDs passed in the second call will be merged with the customer IDs passed in the first call. Is this intentional/desired? If so, maybe take advantage of the `assign` utility we have:

```
assign(customerIds, ids);
```

File: src/components/Identity/index.js [hunk 20]
```diff
@@ -100,6 +148,12 @@ const createIdentity = ({ config, logger, cookie }) => {
     commands: {
       getEcid() {
         return optIn.whenOptedIn().then(getEcid);
+      },
+      setCustomerIds(options) {
+        return optIn.whenOptedIn().then(setCustomerIds(options));
```
Comment (Aaronius) [inline src/components/Identity/index.js]: Shouldn't this be:

```
return optIn.whenOptedIn().then(() => setCustomerIds(options));
```

I'm assuming you're wanting to wait for opt in before `setCustomerIds` is called.

File: src/components/Identity/util.js [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    .slice(0, -1);
+};
+
+const createHashFromString = string => {
+  return string.split('')
+    .map(char => char.charCodeAt(0))
+    .reduce((hash, charCode) => {
+      const shift = (hash << 5) - hash + charCode; // eslint-disable-line no-bitwise
+      return shift & shift; // eslint-disable-line no-bitwise
+    }, 0);
```
Comment (Aaronius) [inline src/components/Identity/util.js]: This hashing algorithm is based on Java's hashCode algorithm, which isn't really a standard. It's probably worth using a better, more standard algorithm. CRC32, MD5, and SHA-1, while not cryptographically secure, are fine for comparing two inputs and would all result in a much smaller hash than SHA-256. While there's no native API for generating a CRC32, it would likely be the fastest, produce the smallest hash, and be the most appropriate for what we're trying to do here. It's probably the smallest amount of code too, even when taking into consideration that SHA-1 has a native API. There's a 1 in 4,294,967,296 chance of collision between two inputs with CRC32, which should be sufficient. I believe this is a proper implementation of CRC32: http://crc32.nichabi.com/javascript-function.php

Comment (Aaronius) [reply src/components/Identity/util.js]: You probably want to run some UTF-8 strings through it and compare the output to some online CRC32 hashing tools.

Comment (Aaronius) [reply src/components/Identity/util.js]: CRC32 returns a number. Since cookies are stored as strings, each digit in the CRC23 number is represented as one character. You can shorten the number of character needed by converting the number to a string that takes advantage of characters other than 0-9. So if you take the output of CRC32 and call `.toString(36)` you'll get a string using characters 0-9 and A-Z, resulting in a string that's about 6 or 7 characters long instead of a 9 or 10 characters. Technically, we could use an even higher radix by leveraging lowercase characters, etc., but there's no native API for that so it's probably not worth the code it would take to implement it.

File: src/components/Identity/index.js [hunk 22]
```diff
@@ -12,28 +12,76 @@ governing permissions and limitations under the License.
...
+    const event = createEvent();
+    Object.keys(ids).forEach(key => {
+      customerIds[key] = ids[key];
+    });
+    const normalizedCustomerIds = normalizeCustomerIds(customerIds);
+    const customerIdsHash = createHashFromString(
+      serializeCustomerIds(normalizedCustomerIds)
+    );
+    const hasSynced = customerIdsHash === cookie.get(CUSTOMER_ID_HASH);
+    event.mergeMeta({ identity: { customerIds, hasSynced } });
```
Comment (Aaronius) [inline src/components/Identity/index.js]: This is probably a dumb question because I don't really understand customer ID syncing, but why do we still send the customer IDs to the server if `hasSynced` is true?

**Comment by jfkhoury** [comment]
@jileeshs Should we incorporate all the comments and start over to clean up this PR?

File: src/components/Identity/identityUtil.js (hunk 12 above)
Comment (jileeshs) [reply src/components/Identity/identityUtil.js]: It is working on IE11, tests are passing

File: src/components/Identity/identityUtil.js (hunk 13 above)
Comment (jileeshs) [reply src/components/Identity/identityUtil.js]: @Aaronius ,
I updated the normalizeCustomerIds method to sort according to id namespaces and then take only id and authState in the order.

File: src/components/Identity/identityUtil.js (hunk 14 above)
Comment (jileeshs) [reply src/components/Identity/identityUtil.js]: Thank you @Aaronius for code sample.
I have updated this PR.

File: src/components/Identity/index.js (hunk 15 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: Done

File: src/components/Identity/index.js (hunk 19 above)
Comment (jileeshs) [reply src/components/Identity/index.js]: Yes, it is the desired behavior. 
I will use the assign utility

File: src/components/Identity/index.js (hunk 6 above)
Comment (raresvla) [reply src/components/Identity/index.js]: @jileeshs If I read this correctly, it looks like you're passing both the (declared) customer IDs and a flag to specify that _nothing changed_ in the same object, under `meta.identity`.

_We need to tweak this, as Konductor expects identities to be sent in the `identityMap`._ Identities can be either set at payloads' root level, or at event level. 

What's interesting to notice further is that this part of context data is deeply merged in Konductor: event-level `identityMap` gets merged with the root level `identityMap` entries; this helps you, for example, set the ECID at root level and then have customerIDs sent in at event level, when they become available to Alloy. This is particularly handy when you pass in a batch of events to Konductor -- if only one event is sent, it does not matter where you declare the identities (root or event level).

To sum up:
1. Declared customer IDs need to be sent in `identityMap`, either as root-level or event-level context data
2. To limit the number of calls we send in to the `IdentityService` downstream of Konductor, we need that extra information in `metadata`. It can be either a flag which says "noting changed" (as it looks like it is now) or the opposite, "a new identity has been declared".

Comment (jfkhoury) [reply src/components/Identity/index.js]: Thank you @raresvla.

Should we go with:

```
metadata: {
    identity: { // Do we need this?
        newDeclaredId: true|false
    }
}
```
@Aaronius @jileeshs

Comment (jfkhoury) [reply src/components/Identity/index.js]: @jileeshs Once we get an answer on this, let's update the PR, close this one and re-open a new one that includes all the new changes from yesterday. Thanks!

Comment (Aaronius) [reply src/components/Identity/index.js]: Maybe `customerIdAdded` or `customerIdChanged` (whichever is more technically accurate) instead of `newDeclaredId`? I feel like `customerIdAdded` might suggest that the value should be a boolean more than `newDeclaredId`. The term "declared" seems new too. If it's used elsewhere or we feel like it helps specify what we're talking about, then that might be good to use, but otherwise I would say we stick to the terms we already have (e.g., "customer ID").

Comment (jfkhoury) [reply src/components/Identity/index.js]: @jileeshs I agree with @Aaronius, let's go with `customerIdChanged` since it's more accurate because an ID is really the ID itself and the Auth State; so the ID can stay the same but the auth state can change.

Comment (jfkhoury) [reply src/components/Identity/index.js]: Thank you @Aaronius!

**Comment by jileeshs** [comment]
This PR has a lot of back and forth messages. I made most of the changes. Closing this one creating a fresh one.

File: src/components/Identity/index.js (hunk 6 above)
Comment (Aaronius) [reply src/components/Identity/index.js]: The ship has probably sailed on this one, but why "customer" and not "user"? The term "customer" seems to suggest an ecommerce context, which doesn't work well with free providers (publishers, social networks, etc). If "customer" is too engrained at this point, just let me know and I'll go on my way...

Comment (jonsnyder) [reply src/components/Identity/index.js]: I always thought of it as in Adobe's Customer's Custom ID.  i.e. the ID that is owned by Adobe's Customer.

Comment (Aaronius) [reply src/components/Identity/index.js]: Oh that's interesting. I never thought of it that way.
